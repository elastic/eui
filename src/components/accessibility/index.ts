export { EuiKeyboardAccessible } from './keyboard_accessible';
export { EuiScreenReaderOnly } from './screen_reader';
export { EuiSkipLink } from './skip_link';
