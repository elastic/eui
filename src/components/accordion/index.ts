export { EuiAccordion, EuiAccordionProps } from './accordion';
