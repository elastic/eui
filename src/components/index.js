export {
  EuiAccordion,
} from './accordion';

export {
  EuiAvatar,
} from './avatar';

export {
  EuiKeyboardAccessible,
  EuiScreenReaderOnly,
} from './accessibility';

export {
  EuiBottomBar,
} from './bottom_bar';

export {
  EuiButton,
  EuiButtonEmpty,
  EuiButtonIcon,
} from './button';

export {
  EuiBadge,
} from './badge';

export {
  EuiCard,
} from './card';

export {
  EuiCallOut,
} from './call_out';

export {
  EuiCode,
  EuiCodeBlock,
  EuiCodeBlockImpl,
} from './code';

export {
  EuiCodeEditor,
} from './code_editor';

export {
  EuiColorPicker,
} from './color_picker';

export {
  EuiContextMenu,
  EuiContextMenuPanel,
  EuiContextMenuItem,
} from './context_menu';

export {
  EuiDelayHide
} from './delay_hide';

export {
  EuiDescriptionList,
  EuiDescriptionListTitle,
  EuiDescriptionListDescription,
} from './description_list';

export {
  EuiErrorBoundary,
} from './error_boundary';

export {
  EuiExpression,
  EuiExpressionButton,
} from './expression';

export {
  EuiFilterButton,
  EuiFilterGroup,
  EuiFilterSelectItem
} from './filter_group';

export {
  EuiFlexGroup,
  EuiFlexGrid,
  EuiFlexItem,
} from './flex';

export {
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutFooter,
  EuiFlyoutHeader,
} from './flyout';

export {
  EuiCheckbox,
  EuiCheckboxGroup,
  EuiFieldNumber,
  EuiFieldPassword,
  EuiFieldSearch,
  EuiFieldText,
  EuiFilePicker,
  EuiForm,
  EuiFormControlLayout,
  EuiFormErrorText,
  EuiFormHelpText,
  EuiFormLabel,
  EuiFormRow,
  EuiRadio,
  EuiRadioGroup,
  EuiRange,
  EuiSelect,
  EuiSwitch,
  EuiTextArea,
  EuiValidatableControl,
} from './form';

export {
  EuiHeader,
  EuiHeaderAlert,
  EuiHeaderBreadcrumb,
  EuiHeaderBreadcrumbCollapsed,
  EuiHeaderBreadcrumbs,
  EuiHeaderLogo,
  EuiHeaderSection,
  EuiHeaderSectionItem,
  EuiHeaderSectionItemButton,
} from './header';

export {
  EuiHealth,
} from './health';

export {
  EuiHorizontalRule,
} from './horizontal_rule';

export {
  ICON_TYPES,
  EuiIcon,
} from './icon';

export {
  EuiImage,
} from './image';

export {
  EuiLoadingKibana,
  EuiLoadingChart,
  EuiLoadingSpinner,
} from './loading';

export {
  EuiKeyPadMenu,
  EuiKeyPadMenuItem,
  EuiKeyPadMenuItemButton,
} from './key_pad_menu';

export {
  EuiLink,
} from './link';

export {
  EUI_MODAL_CANCEL_BUTTON,
  EUI_MODAL_CONFIRM_BUTTON,
  EuiConfirmModal,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
} from './modal';

export {
  EuiOverlayMask,
} from './overlay_mask';

export {
  EuiOutsideClickDetector,
} from './outside_click_detector';

export {
  EuiPage,
  EuiPageBody,
  EuiPageContent,
  EuiPageContentBody,
  EuiPageContentHeader,
  EuiPageContentHeaderSection,
  EuiPageHeader,
  EuiPageHeaderSection,
  EuiPageSideBar,
} from './page';

export {
  EuiPagination,
  EuiPaginationButton,
} from './pagination';

export {
  EuiPanel,
} from './panel';

export {
  EuiPopover,
  EuiPopoverTitle,
} from './popover';

export {
  EuiPortal,
} from './portal';

export {
  EuiProgress,
} from './progress';

export {
  EuiSearchBar
} from './search_bar';

export {
  EuiSpacer,
} from './spacer';

export {
  EuiSteps,
  EuiSubSteps,
  EuiStepsHorizontal,
} from './steps';

export {
  EuiTable,
  EuiTableBody,
  EuiTableHeader,
  EuiTableHeaderButton,
  EuiTableHeaderCell,
  EuiTableHeaderCellCheckbox,
  EuiTablePagination,
  EuiTableRow,
  EuiTableRowCell,
  EuiTableRowCellCheckbox,
} from './table';

export {
  EuiBasicTable,
  EuiInMemoryTable
} from './basic_table';

export {
  EuiTab,
  EuiTabs,
} from './tabs';

export {
  EuiSideNav,
} from './side_nav';

export {
  EuiGlobalToastList,
  EuiGlobalToastListItem,
  EuiToast,
} from './toast';

export {
  EuiIconTip,
  EuiToolTip,
} from './tool_tip';

export {
  EuiTitle,
} from './title';

export {
  EuiText,
  EuiTextColor,
} from './text';
