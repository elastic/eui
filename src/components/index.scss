
// Components

@import 'accessibility/index';
@import 'accordion/index';
@import 'aspect_ratio/index';
@import 'avatar/index';
@import 'badge/index';
@import 'basic_table/index';
@import 'beacon/index';
@import 'bottom_bar/index';
@import 'button/index';
@import 'breadcrumbs/index';
@import 'call_out/index';
@import 'card/index';
@import 'code/index';
@import 'code_editor/index';
@import 'collapsible_nav/index';
@import 'color_picker/index';
@import 'combo_box/index';
@import 'comment_list/index';
@import 'context_menu/index';
@import 'control_bar/index';
@import 'date_picker/index';
@import 'datagrid/index';
@import 'description_list/index';
@import 'drag_and_drop/index';
@import 'empty_prompt/index';
@import 'error_boundary/index';
@import 'expression/index';
@import 'facet/index';
@import 'filter_group/index';
@import 'flex/index';
@import 'flyout/index';
@import 'form/index';
@import 'header/index';
@import 'health/index';
@import 'horizontal_rule/index';
@import 'icon/index';
@import 'image/index';
@import 'key_pad_menu/index';
@import 'link/index';
@import 'list_group/index';
@import 'loading/index';
@import 'mark/index';
@import 'modal/index';
@import 'nav_drawer/index';
@import 'overlay_mask/index';
@import 'page/index';
@import 'pagination/index';
@import 'panel/index';
@import 'popover/index';
@import 'portal/index';
@import 'progress/index';
@import 'tree_view/index';
@import 'resizable_container/index';
@import 'side_nav/index';
@import 'spacer/index';
@import 'search_bar/index';
@import 'selectable/index';
@import 'stat/index';
@import 'steps/index';
@import 'suggest/index';
@import 'table/index';
@import 'tabs/index';
@import 'title/index';
@import 'toast/index';
@import 'token/index';
@import 'toggle/index';
@import 'tool_tip/index';
@import 'tour/index';
@import 'text/index';
