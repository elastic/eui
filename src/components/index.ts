/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

export * from './accessibility';

export * from './accordion';

export * from './aspect_ratio';

export * from './auto_sizer';

export * from './avatar';

export * from './badge';

export * from './beacon';

export * from './bottom_bar';

export * from './breadcrumbs';

export * from './button';

export * from './call_out';

export * from './card';

export * from './code';

export * from './collapsible_nav';
export * from './collapsible_nav_beta';

export * from './color_picker';

export * from './combo_box';

export * from './comment_list';

export * from './context';

export * from './context_menu';

export * from './copy';

export * from './datagrid';

export * from './date_picker';

export * from './delay_hide';
export * from './delay_render';

export * from './description_list';

export * from './drag_and_drop';

export * from './empty_prompt';

export * from './error_boundary';

export * from './expression';

export * from './filter_group';

export * from './facet';

export * from './flex';

export * from './flyout';

export * from './focus_trap';

export * from './form';

export * from './header';

export * from './health';

export * from './highlight';

export * from './horizontal_rule';

export * from './icon';

export * from './image';

export * from './inner_text';

export * from './inline_edit';

export * from './i18n';

export * from './loading';

export * from './key_pad_menu';

export * from './link';

export * from './list_group';

export * from './markdown_editor';

export * from './mark';

export * from './modal';

export * from './observer/mutation_observer';

export * from './outside_click_detector';

export * from './overlay_mask';

export * from './page';

export * from './page_template';

export * from './pagination';

export * from './panel';

export * from './popover';

export * from './portal';

export * from './progress';

export * from './provider';

export * from './tree_view';

export * from './observer/resize_observer';

export * from './search_bar';

export * from './selectable';

export * from './side_nav';

export * from './skeleton';

export * from './spacer';

export * from './stat';

export * from './steps';

export * from './table';

export * from './token';

export * from './tour';

export * from './basic_table';

export * from './tabs';

export * from './text';

export * from './text_diff';

export * from './text_truncate';

export * from './timeline';

export * from './title';

export * from './toast';

export * from './tool_tip';

export * from './responsive';

export * from './resizable_container';

export * from './common';
