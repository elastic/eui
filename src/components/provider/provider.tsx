/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { PropsWithChildren } from 'react';
import type { EmotionCache } from '@emotion/css';

import {
  EuiThemeProvider,
  EuiThemeProviderProps,
  EuiThemeSystem,
  CurrentEuiBreakpointProvider,
} from '../../services';
import { emitEuiProviderWarning } from '../../services/theme/warning';
import { cache as fallbackCache } from '../../services/emotion/css';

import {
  EuiGlobalStyles,
  EuiGlobalStylesProps,
} from '../../global_styling/reset/global_styles';
import { EuiUtilityClasses } from '../../global_styling/utility/utility';
import { EuiThemeAmsterdam } from '../../themes';

import { EuiCacheProvider } from './cache';
import { EuiProviderNestedCheck, useIsNestedEuiProvider } from './nested';
import {
  EuiComponentDefaults,
  EuiComponentDefaultsProvider,
} from './component_defaults';

const isEmotionCacheObject = (
  obj: EmotionCache | Object
): obj is EmotionCache => obj.hasOwnProperty('key');

export interface EuiProviderProps<T>
  extends PropsWithChildren,
    EuiGlobalStylesProps,
    Pick<EuiThemeProviderProps<T>, 'colorMode' | 'modify'> {
  /**
   * Provide a specific EuiTheme; Defaults to EuiThemeAmsterdam;
   * Pass `null` to remove all theming including global reset
   */
  theme?: EuiThemeSystem | null;
  /**
   * Provide global styles via `@emotion/react` `Global` for your custom theme.
   * Pass `false` to remove the default EUI global styles.
   */
  globalStyles?: false | ((params: any) => JSX.Element | null);
  /**
   * Provide utility classes.
   * Pass `false` to remove the default EUI utility classes.
   */
  utilityClasses?: false | ((params: any) => JSX.Element | null);
  /**
   * Provide a cache configuration(s) from `@emotion/cache`.
   *
   * - `default` will encompass all Emotion styles, including consumer defined appliction styles, not handled by nested cache instances.
   * - `global` will scope all EUI global and reset styles.
   * - `utility` will scope all EUI utility class styles.
   *
   * A cache instance provided as the sole value will function the same as the `default` cache.
   */
  cache?:
    | EmotionCache
    | {
        default?: EmotionCache;
        global?: EmotionCache;
        utility?: EmotionCache;
      };
  /**
   * Allows configuring specified component defaults across all usages, overriding
   * baseline EUI component defaults.
   *
   * Not all components will be supported, and configurable component defaults
   * will be considered on a case-by-case basis.
   *
   * Individual component prop usages will always override these defaults.
   */
  componentDefaults?: EuiComponentDefaults;
}

export const EuiProvider = <T extends {} = {}>({
  cache = fallbackCache,
  theme = EuiThemeAmsterdam,
  globalStyles: Globals = EuiGlobalStyles,
  utilityClasses: Utilities = EuiUtilityClasses,
  colorMode,
  modify,
  componentDefaults,
  children,
}: PropsWithChildren<EuiProviderProps<T>>) => {
  const isNested = useIsNestedEuiProvider();
  if (isNested) {
    const providerMessage = `\`EuiProvider\` should not be nested or used more than once, other than at the top level of your app.
    Use \`EuiThemeProvider\` instead for nested component-level theming: https://ela.st/euiprovider.`;

    emitEuiProviderWarning(providerMessage);
    return children as any;
  }

  let defaultCache;
  let globalCache;
  let utilityCache;
  if (cache) {
    if (isEmotionCacheObject(cache)) {
      cache.compat = true;
      defaultCache = cache;
    } else {
      if (cache.default) {
        cache.default.compat = true;
      }
      defaultCache = cache.default;
      if (cache.global) {
        cache.global.compat = true;
      }
      globalCache = cache.global;
      if (cache.utility) {
        cache.utility.compat = true;
      }
      utilityCache = cache.utility;
    }
  }

  return (
    <EuiProviderNestedCheck>
      <EuiCacheProvider cache={defaultCache ?? fallbackCache}>
        <EuiThemeProvider
          theme={theme ?? undefined}
          colorMode={colorMode}
          modify={modify}
        >
          {theme && (
            <>
              <EuiCacheProvider
                cache={globalCache}
                children={Globals && <Globals />}
              />
              <EuiCacheProvider
                cache={utilityCache}
                children={Utilities && <Utilities />}
              />
            </>
          )}
          <EuiComponentDefaultsProvider componentDefaults={componentDefaults}>
            <CurrentEuiBreakpointProvider>
              {children}
            </CurrentEuiBreakpointProvider>
          </EuiComponentDefaultsProvider>
        </EuiThemeProvider>
      </EuiCacheProvider>
    </EuiProviderNestedCheck>
  );
};
