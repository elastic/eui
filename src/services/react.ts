/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { unstable_batchedUpdates } from 'react-dom';

const _queue: Function[] = [];

function processQueue() {
  // the queued functions trigger react setStates which, if unbatched,
  // each cause a full update->render->dom pass _per function_
  // instead, tell React to wait until all updates are finished before re-rendering
  unstable_batchedUpdates(() => {
    for (let i = 0; i < _queue.length; i++) {
      _queue[i]();
    }
    _queue.length = 0;
  });
}

export function enqueueStateChange(fn: Function) {
  if (_queue.length === 0) {
    setTimeout(processQueue);
  }
  _queue.push(fn);
}
