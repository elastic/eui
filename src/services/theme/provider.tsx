/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  useContext,
  useEffect,
  useRef,
  useState,
  PropsWithChildren,
} from 'react';
import isEqual from 'lodash/isEqual';

import {
  EuiSystemContext,
  EuiThemeContext,
  EuiModificationsContext,
  EuiColorModeContext,
} from './context';
import { buildTheme, getColorMode, getComputed, mergeDeep } from './utils';
import {
  EuiThemeColorMode,
  EuiThemeSystem,
  EuiThemeModifications,
} from './types';

export interface EuiThemeProviderProps<T> {
  theme?: EuiThemeSystem<T>;
  colorMode?: EuiThemeColorMode;
  modify?: EuiThemeModifications<T>;
  children: any;
}

export function EuiThemeProvider<T = {}>({
  theme: _system,
  colorMode: _colorMode,
  modify: _modifications,
  children,
}: PropsWithChildren<EuiThemeProviderProps<T>>) {
  const parentSystem = useContext(EuiSystemContext);
  const parentModifications = useContext(EuiModificationsContext);
  const parentColorMode = useContext(EuiColorModeContext);
  const parentTheme = useContext(EuiThemeContext);

  const [system, setSystem] = useState(_system || parentSystem);
  const prevSystemKey = useRef(system.key);

  const [modifications, setModifications] = useState<EuiThemeModifications>(
    mergeDeep(parentModifications, _modifications)
  );
  const prevModifications = useRef(modifications);

  const [colorMode, setColorMode] = useState<EuiThemeColorMode>(
    getColorMode(_colorMode, parentColorMode)
  );
  const prevColorMode = useRef(colorMode);

  // TODO: Flip if return to using parent
  const isParentTheme = useRef(
    prevSystemKey.current === parentSystem.key &&
      colorMode === parentColorMode &&
      isEqual(parentModifications, modifications)
  );

  const [theme, setTheme] = useState(
    isParentTheme.current && Object.keys(parentTheme).length
      ? parentTheme
      : getComputed(
          system,
          buildTheme(modifications, `_${system.key}`) as typeof system,
          colorMode
        )
  );

  useEffect(() => {
    const newSystem = _system || parentSystem;
    if (prevSystemKey.current !== newSystem.key) {
      setSystem(newSystem);
      prevSystemKey.current = newSystem.key;
      isParentTheme.current = false;
    }
  }, [_system, parentSystem]);

  useEffect(() => {
    const newModifications = mergeDeep(parentModifications, _modifications);
    if (!isEqual(prevModifications.current, newModifications)) {
      setModifications(newModifications);
      prevModifications.current = newModifications;
      isParentTheme.current = false;
    }
  }, [_modifications, parentModifications]);

  useEffect(() => {
    const newColorMode = getColorMode(_colorMode, parentColorMode);
    if (!isEqual(newColorMode, prevColorMode.current)) {
      setColorMode(newColorMode);
      prevColorMode.current = newColorMode;
      isParentTheme.current = false;
    }
  }, [_colorMode, parentColorMode]);

  useEffect(() => {
    if (!isParentTheme.current) {
      setTheme(
        getComputed(
          system,
          buildTheme(modifications, `_${system.key}`) as typeof system,
          colorMode
        )
      );
    }
  }, [colorMode, system, modifications]);

  return (
    <EuiColorModeContext.Provider value={colorMode}>
      <EuiSystemContext.Provider value={system}>
        <EuiModificationsContext.Provider value={modifications}>
          <EuiThemeContext.Provider value={theme}>
            {children}
          </EuiThemeContext.Provider>
        </EuiModificationsContext.Provider>
      </EuiSystemContext.Provider>
    </EuiColorModeContext.Provider>
  );
}
