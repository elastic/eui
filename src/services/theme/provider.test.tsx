/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { render } from '@testing-library/react'; // Note - don't use the EUI custom RTL `render`, as it auto-wraps an `EuiProvider`
import { css } from '@emotion/react';

import { EuiProvider } from '../../components/provider';
import { EuiThemeProvider } from './provider';

describe('EuiThemeProvider', () => {
  describe('colorMode', () => {
    it('sets `colorMode`', () => {
      const { getByText } = render(
        <>
          <EuiThemeProvider colorMode="light">
            <div css={({ euiTheme }) => ({ color: euiTheme.colors.fullShade })}>
              Light mode
            </div>
            <EuiThemeProvider colorMode="inverse">
              <div
                css={({ euiTheme }) => ({ color: euiTheme.colors.fullShade })}
              >
                Inverse of light mode
              </div>
            </EuiThemeProvider>
          </EuiThemeProvider>
          <EuiThemeProvider colorMode="dark">
            <div css={({ euiTheme }) => ({ color: euiTheme.colors.fullShade })}>
              Dark mode
            </div>
            <EuiThemeProvider colorMode="inverse">
              <div
                css={({ euiTheme }) => ({ color: euiTheme.colors.fullShade })}
              >
                Inverse of dark mode
              </div>
            </EuiThemeProvider>
          </EuiThemeProvider>
        </>
      );

      expect(getByText('Light mode')).toHaveStyleRule('color', '#000');
      expect(getByText('Dark mode')).toHaveStyleRule('color', '#FFF');
      expect(getByText('Inverse of light mode')).toHaveStyleRule(
        'color',
        '#FFF'
      );
      expect(getByText('Inverse of dark mode')).toHaveStyleRule(
        'color',
        '#000'
      );
    });
  });

  describe('modify', () => {
    it('allows overriding theme tokens', () => {
      const { getByText } = render(
        <EuiThemeProvider
          modify={{ colors: { LIGHT: { primary: 'hotpink' } } }}
        >
          <div css={({ euiTheme }) => ({ color: euiTheme.colors.primary })}>
            Modified
          </div>
        </EuiThemeProvider>
      );

      expect(getByText('Modified')).toHaveStyleRule('color', 'hotpink');
    });
  });

  describe('nested EuiThemeProviders', () => {
    it('renders with a span wrapper that sets the inherited text color', () => {
      const { container } = render(
        <EuiProvider>
          Top-level provider{' '}
          <EuiThemeProvider colorMode="dark">
            Nested
            <EuiThemeProvider colorMode="light">
              Double nested
              <EuiThemeProvider colorMode="inverse">
                Triple nested
              </EuiThemeProvider>
            </EuiThemeProvider>
          </EuiThemeProvider>
        </EuiProvider>
      );

      expect(container).toMatchSnapshot();
    });

    it('allows customizing the span wrapper with `wrapperProps`', () => {
      const customCss = css`
        display: flex;
      `;

      const { container } = render(
        <EuiProvider>
          Top-level provider{' '}
          <EuiThemeProvider
            colorMode="dark"
            wrapperProps={{
              className: 'test',
              'data-test-subj': 'nested',
              css: customCss,
            }}
          >
            Nested
          </EuiThemeProvider>
        </EuiProvider>
      );

      expect(container).toMatchSnapshot();
      expect(container.querySelector('.test')).toBeTruthy();
    });

    it('allows avoiding the extra span wrapper with `wrapperProps.cloneElement`', () => {
      const { container } = render(
        <EuiProvider>
          Top-level provider{' '}
          <EuiThemeProvider
            colorMode="dark"
            wrapperProps={{ cloneElement: true, className: 'hello' }}
          >
            <div className="world">clone provider color onto div</div>
          </EuiThemeProvider>
        </EuiProvider>
      );

      expect(container).toMatchSnapshot();
      expect(container.querySelector('.hello.world')).toBeTruthy();
    });
  });
});
