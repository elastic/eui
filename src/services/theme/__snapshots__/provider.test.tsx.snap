// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiThemeProvider nested EuiThemeProviders allows avoiding the extra span wrapper with \`wrapperProps.cloneElement\` 1`] = `
<div>
  Top-level provider
   
  <div
    class="world hello emotion-euiColorMode-dark-colorClassName"
  >
    clone provider color onto div
  </div>
</div>
`;

exports[`EuiThemeProvider nested EuiThemeProviders allows customizing the span wrapper with \`wrapperProps\` 1`] = `
<div>
  Top-level provider
   
  <span
    class="euiThemeProvider test css-1mg4zk6-customCss-euiColorMode-dark-colorClassName"
    data-test-subj="nested"
  >
    Nested
  </span>
</div>
`;

exports[`EuiThemeProvider nested EuiThemeProviders renders with a span wrapper that sets the inherited text color 1`] = `
<div>
  Top-level provider
   
  <span
    class="euiThemeProvider emotion-euiColorMode-dark-colorClassName"
  >
    Nested
    <span
      class="euiThemeProvider emotion-euiColorMode-light-colorClassName"
    >
      Double nested
      <span
        class="euiThemeProvider emotion-euiColorMode-inverse-colorClassName"
      >
        Triple nested
      </span>
    </span>
  </span>
</div>
`;
